### PR TITLE
Gemma 4 VLM: fuse float32 logit softcapping

### PR DESCRIPTION
**Pull request by aleroot**
## Proposed changes

Fuse `tanh(logits / cap) * cap` in the Gemma 4 VLM text decoder using a stateless compiled function. Pass the cap as a runtime input and explicitly promote float16/bfloat16 logits to float32 before compilation to preserve the original rounding.

This changes one production call site. Public APIs, checkpoint loading, cache handling, and the text-only LLM implementation remain unchanged.

## Performance

Apple M2, Release build, bfloat16 logits with a float32 cap. Each run alternated the original and fused expressions within the same process, measuring warmed, synchronized operation latency.

| Run | Original | Fused | Latency reduction |
|---|---:|---:|---:|
| 1 | 309.8 µs | 298.3 µs | 3.7% |
| 2 | 316.8 µs | 311.2 µs | 1.8% |
| 3 | 318.1 µs | 281.0 µs | 11.7% |

**Median latency reduction: 3.7% for softcapping.** No reliable whole-model tokens/sec improvement was established.

## Checklist

Put an `x` in the boxes that apply.

- [x] I have read the [CONTRIBUTING](https://github.com/ml-explore/mlx-swift-lm/blob/main/CONTRIBUTING.md) document
- [x] I have run `pre-commit run --all-files` to format my code / installed pre-commit prior to committing changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have updated the necessary documentation (if needed)

## AI usage

- [ ] I have read this PR description in full and approve it as my own, and it
      accurately describes the code changes.
- AI usage disclosure:

### CODE REVIEW

File: Libraries/MLXLMCommon/Gemma4Fusions.swift [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+
+private let compiledGemma4Softcap = compile(shapeless: true) {
+    (logits: MLXArray, cap: MLXArray) in tanh(logits / cap) * cap
+}
+
+package func gemma4LogitSoftcap(_ logits: MLXArray, _ cap: MLXArray) -> MLXArray {
+    guard cap.dtype == .float32 else { return tanh(logits / cap) * cap }
+    // Promote before tracing: mixed float16/float32 division can round differently inside compile.
+    if logits.dtype == .float16 || logits.dtype == .bfloat16 {
+        return compiledGemma4Softcap(logits.asType(.float32), cap)
```
Comment (davidkoski) [inline Libraries/MLXLMCommon/Gemma4Fusions.swift:11]: I wonder if there should be a variant that does asType inside the compiled code?  Otherwise this might  materialize a buffer with the asType.

Comment (davidkoski) [reply Libraries/MLXLMCommon/Gemma4Fusions.swift:11]: Something like this:

```swift
private let compiledGemma4SoftcapF16 = compile(shapeless: true) {
    (logits: MLXArray, cap: MLXArray) in tanh(logits.asType(.float32) / cap) * cap
}
```

Here is what I get (m5 max) with the code as-is:

[Gemma4 fusion] VLM softcap: original 222.2881249326747 µs, fused 204.83645843341947 µs
[Gemma4 fusion] VLM softcap: original 216.4189582981635 µs, fused 205.74854162987322 µs
[Gemma4 fusion] VLM softcap: original 214.68583334353752 µs, fused 201.6464582993649 µs
[Gemma4 fusion] VLM softcap: original 219.23500011325814 µs, fused 208.5514581995085 µs
[Gemma4 fusion] VLM softcap: original 221.02916656876914 µs, fused 209.082291694358 µs

and 

[Gemma4 fusion] VLM softcap: original 220.65687502617948 µs, fused 201.9302082771901 µs
[Gemma4 fusion] VLM softcap: original 220.3681250102818 µs, fused 195.95999998273328 µs
[Gemma4 fusion] VLM softcap: original 225.98395837121643 µs, fused 195.03729170537554 µs
[Gemma4 fusion] VLM softcap: original 224.58708335761912 µs, fused 202.8235416219104 µs
[Gemma4 fusion] VLM softcap: original 228.2270832802169 µs, fused 199.48958346503787 µs

See what you think.

**Review by davidkoski — CHANGES_REQUESTED** [review]
Looks good, see what you think on my comment.  Ready to merge either way.

File: Libraries/MLXLMCommon/Gemma4Fusions.swift (hunk 1 above)
Comment (aleroot) [reply Libraries/MLXLMCommon/Gemma4Fusions.swift:11]: Thanks @davidkoski   I tried it, but it exposes an MLX kernel-cache collision when float16 and bfloat16 run in the same process, producing incorrect outputs. I’d keep the cast outside compilation until the cache-key issue is fixed.

Comment (davidkoski) [reply Libraries/MLXLMCommon/Gemma4Fusions.swift:11]: Oh ouch!

**Review by davidkoski — APPROVED** [review]
